text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,336 @@
+
+1. Ideally, the dependencies, `assert`, `cloudformation-diff` and `cfnspec`
+   are all available to be consumed as regular dependencies but are not,
+   since they are part of the same monorepo.
+
+   Instead, the design 'vendors in' these dependencies, i.e., copies over
+   the source code and rewrites imports, thereby increasing tech debt.
+
+2. Any new feature or bug fix to the new assert module that require changes
+   to the API needs to be performed both in the old assert module and new.
+   This can lead to weird/complicated implementation, since they both follow
+   different design patterns.
+
+3. The 'jest' friendly interfaces that were previously available are no longer
+   available.
+   Users previously using -
+
+   ```ts
+   expect(stack).toHaveResource('Foo::Bar', { ... });
+   ```
+
+   will now have to use
+
+   ```ts
+   const assertions = TemplateAssertions.fromStack(stack);
+   assertions.assertResource('Foo::Bar', { ... });
+   ```
+
+### What are the mitigations and/or long term plans to address these concerns?
+
+Once this module reaches general availability, we will deprecate the old 'assert'
+module. At that point, all of its source code can now be made as part of the new
+'assert' module.
+
+The dependencies on 'cloudformation-diff' and 'cfnspec' will need to be carefully
+evaluated and a decision made on whether to lift-and-shift the relevant code, or
+to move these out of the monorepo as independent modules.
+
+Friendly interfaces for popular testing frameworks such as jest, nunit, junit, etc.
+can be built as separate modules specific to those languages (not jsii modules)
+on top of the new 'assert' library.
+If necessary, this can be owned and maintained by the CDK team, but probably more
+suited for the community to pick up and build as third party modules.
+
+### What alternative solutions did you consider?
+
+See [Appendix B](#appendix-b---alternatives) for alternatives.
+
+### What is the high level implementation plan?
+
+The implementation plan is fairly simple.
+
+- Implement and release the new assert module.
+- Migrate modules to use the new assert module.
+
+### Are there any open issues that need to be addressed later?
+
+None.
+
+## Appendix A - Detailed Design
+
+### Key Requirement
+
+The **key requirement** for the design choices is *dogfooding*.
+The new assert APIs must be usable from within the AWS CDK, and over time,
+all CDK modules that use the old assert module should be migrated to the new one.
+
+This is crucial to keep the assert library active and maintained.
+This is reinforced by the old assert library which, over time, has evolved with
+new features added to it.
+
+### Ship Vehicle
+
+The assert library works on `Stack` which is a construct available in `@aws-cdk/core`
+(or `aws-cdk-lib` in v2). Hence, takes a dependency on this construct.
+
+Given this and working backwards from the key requirement, the new assert module
+must be part of the AWS CDK monorepo.
```

Comment:
This logic doesn't convince me.

Why can't this be a separate module that depends on `aws-cdk-lib`?

Reply:
How would you satisfy the requirement I've identified (in the above section) if this is a separate module?
Do you disagree with that requirement?

Reply:
We can ship a separate experimental assert library from a module in the monorepo, re-writing it as needed at build time.

Same way as we re-write the stable modules today to package them as part of monocdk, and the same way as we plan to ship the existing experimental modules like `@aws-cdk/aws-appmesh` (which has version `1.107.0`) as new modules `@aws-cdk-lib-experiments/aws-appmesh` (with version `0.1.0`).

Reply:
I'll point you again towards the requirements section and quote it here, because your reply seems to ignore it -

> The **key requirement** for the design choices is *dogfooding*.
> The new assert APIs must be usable from within the AWS CDK, and over time,
> all CDK modules that use the old assert module should be migrated to the new one.

If we ship this as an experimental module, I believe we cannot take declare a dependency on it from any of our stable modules.

Reply:
> If we ship this as an experimental module, I believe we cannot take declare a dependency on it from any of our stable modules.

Pretty sure we can, because it will be a `devDependency`.

Reply:
Wait, actually. Why do we need to declare a dependency on this module at all? `aws-cdk-lib` does not ship with any tests...?

Reply:
Maybe but why. We'll still get into weird situations that we don't foresee.

For instance:
let's say, `@aws-cdk/aws-lambda` -> `aws-cdk-assertions` as a dev dependency. What will `aws-cdk-assertions` depend on? `@aws-cdk/core` or `aws-cdk-lib`?
It'll have to be the former, which means v2 customers can't use the new library. Or we'll need to do some re-packaging and re-publishing, etc.

Why deal with all this complexity? If we just publish this as any CDK construct library, it keeps things simple and gets the job done. We can deal with it the same way we handle everything else.
Why do you not like shipping this as any other construct library? The only difference I see is that these constructs does not participate in synthesis, which is acceptable.

I'd rather not create another special module that operates differently from the rest.

Reply:
> Why deal with all this complexity? If we just publish this as any CDK construct library, it keeps things simple and gets the job done. We can deal with it the same way we handle everything else.
Why do you not like shipping this as any other module? The only difference I see is that these does not participate in synthesis, which is acceptable.

Because it bloats the `aws-cdk-lib` module with test code, and forces you to use the ugly `Beta1` names.

I honestly don't see the problem here. We will have a module in the monorepo, `@aws-cdk/assertions` or something, which the single (v1) modules will depend on. This takes care of dogfooding. Then we re-package the module to be released with whatever name we want, `@aws-cdk-lib/assertions` or something else, in version `0.x.y`. It will `peerDepend` on `aws-cdk-lib`.

This is exactly the process we want to have for other experimental modules.

Reply:
Ok. Thanks for the feedback.

I'll take that into consideration before my final decision.

Reply:
@skinny85 I am not sure I understand your arguments against including this is the standard library (besides module size, which is not a very strong argument given this is likely to add a few KiBs).

Reply:
It also allows us to skip the ugly `Beta1` suffixes.

Reply:
That's a fair argument. 
If we can release this library as an experimental module without adding any further complexity, we can do so.

However, after graduating to stable, the module will be included in `aws-cdk-lib` (just like any other experimental module).